service/featuregate/README.md

```diff
@@ -0,0 +1,111 @@
+# Collector Feature Gates
+
+This package provides a mechanism that allows operators to enable and disable
+experimental or transitional features at deployment time. These flags should
+be able to govern the behavior of the application starting as early as possible
+and should be available to every component such that decisions may be made
+based on flags at the component level.
+
+## Usage
+
+Feature gates must be defined and registered with the global registry in
+an `init()` function.  This makes the `Gate` available to be configured and 
+queried with a default value of its `Enabled` property.
+
+```go
+var fancyNewFeatureGate = featuregate.Gate{
+	ID:          "namespaced.uniqueIdentifier",
+	Description: "A brief description of what the gate controls",
+	Enabled:     false,
+}
+
+func init() {
```

Comment:
I think we should recommend that the `init` function lives in the same package as the "factory" for the component that register the gate, correct? This way we guarantee that it does the correct thing.

Also another idea is to have an optional "interface" on the `component.Factory` that factories can implement:
 ```
type component.Gateable interface {
   func Gates() ([]Gate, error)
}
```

This way "Service" becomes the owner of the "Gates" and we can remove the need of the global variable and the static initialization.

Reply:
> I think we should recommend that the `init` function lives in the same package as the "factory" for the component that register the gate, correct? This way we guarantee that it does the correct thing.

Yes, I think the recommendation should be that the `Gate` is registered by the package that will be primarily responsible for acting on it or providing the feature it controls.  That's not always going to be a component with a factory, though.  See the telemetry initialization for an example.  
 
> Also another idea is to have an optional "interface" on the `component.Factory` that factories can implement:
> 
> ```
> type component.Gateable interface {
>   func Gates() ([]Gate, error)
> }
> ```

I think this would be too late.  These gates wouldn't be registered until after the component had been loaded which means that the component can't make use of gates to control initialization.  I think it would also complicate any attempt to provide a list of gates via CLI, either with a flag or a subcommand.

> This way "Service" becomes the owner of the "Gates" and we can remove the need of the global variable and the static initialization.

Do you mean to remove the global feature gate registry?  I think that would put too much responsibility in the `service`, which already feels like it is doing a lot.

Reply:
> I think this would be too late. These gates wouldn't be registered until after the component had been loaded which means that the component can't make use of gates to control initialization. I think it would also complicate any attempt to provide a list of gates via CLI, either with a flag or a subcommand.

Why? We pass the gates at ComponentCreate anyway, so already a limitation that cannot be used until that point.

Reply:
A gate used by a component can't be used until the component is initialized, but that is different than the gate not being available at all until the component is initialized.  How would a gate that affects component initialization be implemented?  How about a gate that affects service initialization or configuration handling?

I feel that overall the changes made in this pass at the PR to separate flags and configuration push feature gate configuration to be too late to be effective for some potential uses.  In the prior implementation where CLI flags directly affected the state of a global feature gate registry it would have been possible to use a feature gate to select between alternate configuration processing methods.  That's no longer possible since the user-selected state isn't made available until after configuration processing has completed.

If gate definition were deferred until component initialization and every gate had to be owned by a component then I'm not sure how the OTel/OpenCensus gate would be implemented.  That isn't part of any component.

Reply:
It is part of the service so can happen, I think the whole complication comes from the fact that we allow configuring via Config if the intention is to be available immediately. If we allow configuring this via the config file the cannot be available before that happens.

For the OC/Otel example, we are working to change the Service to own the telemetry, so that will happen after the configuration is parsed and during Service initialization which is ok.